### PR TITLE
leveldb: implement concurrent compaction

### PR DESCRIPTION
**Pull request by rjl493456442**
This PR runs the isolated compaction concurrently to raise the system performance.

### CODE REVIEW

**Comment by rjl493456442** [comment]
A few more information after running the write benchmarking.

### Benchmarking setting

- Key size 100b
- Value size 100b
- Total value size 10GB
- Write buffer 100MB

### Master

- Database statistic

```
{38290 15m11.250540464s false 0 0 96677349193 66976108702 0 4 [320361511 2492915337 12383826915] 
[3 1481 6045] [0 29072592687 37897539069] [15270499841 29047698903 37849036775] [1m56.873115244s 8m39.370564199s 11m49.064456104s] 143 16 5618 0}
```

- Performance statistic

```
-- batch-100kb.2020-08-06-18:35:19 (17476 events) total time: 1264.8400s
 total size: 10737254400 bytes
  mean mb/s: 8.096 (+- 12.743)
```

### PR 

- Database statistic 

```
{42624 8m53.29404969s false 0 0 119958158300 90251417268 0 33 [427150354 1210462062 1680973110 11882306866] 
[4 623 975 6111] [0 17962660959 27789895644 44474303957] [15270499841 17954901062 27777240464 44425111585] 
[2m33.999790202s 10m35.894896992s 34m55.40700714s 1h0m21.973738336s] 143 20 10243 0}
```

- Performance statistic

```
-- batch-100kb.2020-08-06-19:04:43 (17476 events) total time: 981.0566s
 total size: 10737254400 bytes
  mean mb/s: 10.438 (+- 12.787)
```

File: leveldb/db_compaction.go [hunk 1]
```diff
@@ -786,80 +850,324 @@ func (db *DB) mCompaction() {
...
+			ctx.noseek = true
+			return
+		}
+	}
+	ctx.add(c)
+	wg.Add(1)
+
+	go func() {
+		// Catch the panic in its own goroutine.
+		defer func() {
```
Comment (rjl493456442) [inline leveldb/db_compaction.go:1157]: We catch the panic in the goroutine. In the original implementation, the panic will be passed to the main thread. The behaviour is changed...

**Comment by rjl493456442** [comment]
### Benchmarking setting 2

- Key size 100b
- Value size 100b
- Total value size **20GB**
- Write buffer 100MB

### Master

- Database statistic

```
{74923 51m55.158298393s false 0 0 217969445106 158587878647 0 6 [1281435560 2343980577 12772278361 13972973946] [12 1380 6216 6911] [0 57512431061 101038735425 25313162] [30541031041 57460130601 100920691029 25295421] [3m57.212215565s 21m38.221922474s 39m40.466237826s 598.812645ms] 286 27 11667 0}
```

- Performance statistic

```
-- batch-100kb.2020-08-06-20:48:44 (34952 events) total time: 3845.0401s
 total size: 21474508800 bytes
  mean mb/s: 5.326 (+- 11.064)
```

### PR 

- Database statistic 

```
{79857 32m41.074744105s false 0 0 285748221868 226365733290 0 3 [1067864748 855142281 2348273262 16024399610 10071344883] [10 450 1421 8411 5194] [0 35578820817 56449218779 134285576309 52117385] [30541031041 35561629964 56415397892 134162611693 52087484] [5m14.606409304s 33m38.251109674s 1h22m49.8935168s 3h28m28.534348668s 3.998871177s] 286 31 24112 0}
```

- Performance statistic

```
-- batch-100kb.2020-08-06-19:58:29 (34952 events) total time: 2723.2274s
 total size: 21474508800 bytes
  mean mb/s: 7.520 (+- 12.231)
```

**Comment by syndtr** [comment]
Hi @rjl493456442,

I did quick stress test and this seems to cause some data inconsistency.

```
2020/08/16 19:34:02 FATAL: [02] READER #31.24 K474 snap.Get: leveldb: not found
k1: 0201dde13fa10e61f346b0b2dcb1b286a1735eb555880100007bc4c2780201dde13fa10e61f346b0b2dcb1b286a1735eb555880100007bc4c2784fd8259bb3
 -> k2: 0202409476998cd27fb4ed61d9e9099aae7b2916c71f0000003a2d00990202409476998cd27fb4ed61d9e9099aae7b2916c71f0000003a2d00995347a528f7
2020/08/16 19:34:02 [02] READER #31.24 DONE Snap=leveldb.Snapshot{6871984} Alive=1 IterLatency=1.627µs GetLatency=22.355µs
2020/08/16 19:34:02 [05] READER #24.367 DONE Snap=leveldb.Snapshot{6422524} Alive=1 IterLatency=1.251µs GetLatency=17.331µs
2020/08/16 19:34:02 [03] READER #25.341 DONE Snap=leveldb.Snapshot{6288541} Alive=1 IterLatency=1.273µs GetLatency=20.634µs
2020/08/16 19:34:02 [01] READER #26.280 DONE Snap=leveldb.Snapshot{6418609} Alive=1 IterLatency=1.204µs GetLatency=18.291µs
2020/08/16 19:34:02 [06] READER #15.2 DONE Snap=leveldb.Snapshot{6808439} Alive=1 IterLatency=2.619µs GetLatency=43.109µs
2020/08/16 19:34:02 [02] READER #30.667 DONE Snap=leveldb.Snapshot{6423586} Alive=0 IterLatency=1.249µs GetLatency=18.231µs
2020/08/16 19:34:02 [03] READER #26.85 DONE Snap=leveldb.Snapshot{6736953} Alive=0 IterLatency=1.511µs GetLatency=31.515µs
2020/08/16 19:34:02 [05] READER #25.36 DONE Snap=leveldb.Snapshot{6870922} Alive=0 IterLatency=1.535µs GetLatency=9.862µs
2020/08/16 19:34:02 [01] READER #27.20 DONE Snap=leveldb.Snapshot{6867009} Alive=0 IterLatency=1.517µs GetLatency=24.941µs
2020/08/16 19:34:02 [04] READER #19.3 DONE Snap=leveldb.Snapshot{6864345} Alive=1 IterLatency=1.778µs GetLatency=27.333µs
2020/08/16 19:34:02 [04] READER #18.39 DONE Snap=leveldb.Snapshot{6414921} Alive=0 IterLatency=1.292µs GetLatency=18.948µs
2020/08/16 19:34:02 [04] SCANNER #16 Deleted=9562 Time=4.125932536s
2020/08/16 19:34:02 [04] SCANNER DONE #17
2020/08/16 19:34:02 [03] SCANNER #17 Deleted=1236 Time=4.13560786s
2020/08/16 19:34:02 [03] SCANNER DONE #18
2020/08/16 19:34:02 > Write using transaction
2020/08/16 19:34:02 [06] READER #14.11 DONE Snap=leveldb.Snapshot{6360027} Alive=0 IterLatency=1.673µs GetLatency=21.88µs
2020/08/16 19:34:02 [00] READER #13.3 DONE Snap=leveldb.Snapshot{6275273} Alive=1 IterLatency=1.707µs GetLatency=33.257µs
2020/08/16 19:34:02 [03] WRITER DONE #28
2020/08/16 19:34:02 [03] READER #27.0 DONE Snap=leveldb.Snapshot{6985340} Alive=0 IterLatency=0s GetLatency=0s
2020/08/16 19:34:03 [00] READER #14.2 DONE Snap=leveldb.Snapshot{6723692} Alive=0 IterLatency=1.561µs GetLatency=21.968µs
panic: negative ref: 3933

goroutine 8 [running]:
github.com/syndtr/goleveldb/leveldb.(*session).refLoop.func1(0xf5d, 0xffffffffffffffff, 0xf61)
	/home/syndtr/pg/githubproj/goleveldb/leveldb/session_util.go:89 +0x160
github.com/syndtr/goleveldb/leveldb.(*session).refLoop.func3(0xc005dadb80)
	/home/syndtr/pg/githubproj/goleveldb/leveldb/session_util.go:107 +0xe8
github.com/syndtr/goleveldb/leveldb.(*session).refLoop.func4()
	/home/syndtr/pg/githubproj/goleveldb/leveldb/session_util.go:176 +0x3a9
github.com/syndtr/goleveldb/leveldb.(*session).refLoop(0xc0000d26c0)
	/home/syndtr/pg/githubproj/goleveldb/leveldb/session_util.go:187 +0x44b
created by github.com/syndtr/goleveldb/leveldb.newSession
	/home/syndtr/pg/githubproj/goleveldb/leveldb/session.go:95 +0x2b4
exit status 2
```

I din't dig into the code, but if you do multi-level concurrent compaction you need to take extra precaution.

---

But, I've decided not to merge this or any non-trivial PR, since I most likely to don't have time and energy to test and validate those. 
I'm very sorry about that. Feel free to fork the project though.
Thanks for the effort Gary, appreciate it :)

**Comment by qianbin** [comment]
Hi Gary, I'm also working on concurrent compaction. Could you please share your bench code ?

**Comment by fxfactorial** [comment]
@syndtr possible to take a look at the PR again?

**Pull request closed without merging**